orangecontrib/prototypes/explanation/explainer.py

```diff
@@ -0,0 +1,568 @@
+    of lists with the np.ndarray for each class, when explaining regression,
+    the result is the list of one np.ndarrays.
+    """
+    if isinstance(shap_values[0], np.ndarray):
+        # regression
+        return [np.vstack(shap_values)]
+    else:
+        # classification
+        return [np.vstack(s) for s in zip(*shap_values)]
+
+
+def _explain_trees(
+    model: Model, transformed_data: Table, progress_callback: Callable
+) -> Tuple[
+    Optional[List[np.ndarray]], Optional[np.ndarray], Optional[np.ndarray]
+]:
+    """
+    Computes and returns SHAP values for learners that are explained by
+    TreeExplainer: all sci-kit models based on trees. In case that explanation
+    with TreeExplainer is not possible it returns None
+    """
+    if sparse.issparse(transformed_data.X):
+        # sparse not supported by TreeExplainer, KernelExplainer can handle it
+        return None, None, None
+    try:
+        explainer = TreeExplainer(
+            model.skl_model, feature_perturbation="tree_path_dependent"
+        )
+    except (SHAPError, AttributeError):
+        return None, None, None
+
+    # TreeExplaner cannot explain in normal time more cases than 10000
+    data_sample, sample_mask = _subsample_data(transformed_data, 10000)
+    num_classes = (
+        len(transformed_data.domain.class_var.values)
+        if transformed_data.domain.class_var.is_discrete
+        else None
+    )
+
+    # this method will work in batches since explaining only one attribute
+    # at time the processing timed doubles comparing to batch size 10
+    shap_values = []
+    batch_size = 1  # currently set to 1 to minimize widget blocking
+    for i in range(0, len(data_sample), batch_size):
+        progress_callback(i / len(data_sample))
+        batch = data_sample.X[i : i + batch_size]
+        shap_values.append(
+            explainer.shap_values(batch, check_additivity=False)
+        )
+
+    shap_values = _join_shap_values(shap_values)
+    base_value = explainer.expected_value
+    # when in training phase one class value was missing skl_model do not
+    # output probability for it. For other models it is handled by Orange
+    if num_classes is not None:
+        missing_d = num_classes - len(shap_values)
+        shap_values += [
+            np.zeros(shap_values[0].shape) for _ in range(missing_d)
+        ]
+        base_value = np.hstack((base_value, np.zeros(missing_d)))
+
+    return shap_values, sample_mask, base_value
+
+
+def _explain_other_models(
+    model: Model,
+    transformed_data: Table,
+    transformed_reference_data: Table,
+    progress_callback: Callable,
+) -> Tuple[List[np.ndarray], np.ndarray, np.ndarray]:
+    """
+    Computes SHAP values for any learner with KernelExplainer.
+    """
+    # 1000 is a number that for normal data and model do not take so long
+    data_sample, sample_mask = _subsample_data(transformed_data, 1000)
+
+    try:
+        ref = kmeans(transformed_reference_data.X, k=10)
```

Comment:
Does this work for data being smaller than 10?

Reply:
It works. It raises a ValueError which is caught and then sample is used instead of k-means

Reply:
I added a test for such case